### PR TITLE
WT-12818 No need to copy the key if we already set the commit timestamp

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

File: src/include/txn_inline.h [hunk 1]
```diff
@@ -115,8 +115,13 @@ __wt_txn_op_set_key(WT_SESSION_IMPL *session, const WT_ITEM *key)
 
     op = txn->mod + txn->mod_count - 1;
 
-    if (WT_SESSION_IS_CHECKPOINT(session) || WT_IS_HS(op->btree->dhandle) ||
-      WT_IS_METADATA(op->btree->dhandle) || F_ISSET(txn, WT_TXN_AUTOCOMMIT))
+    /*
+     * We save the key for resolving the prepared updates. However, if we have already set the
+     * commit timestamp, the transaction cannot be prepared. Therefore, no need to save the key.
+     */
+    if (F_ISSET(txn, WT_TXN_HAS_TS_COMMIT) || WT_SESSION_IS_CHECKPOINT(session) ||
```
Comment (luke-pearson) [inline src/include/txn_inline.h:122]: This if statement has become so complex it almost deserves it's own function `__wt_txn_op_need_save_key`. Whether that is called from within it or externally I don't really mind. But something here feels like it needs to be refactored. The function itself is called `set_key` not `set_key_if_blahblahblah`.

File: src/include/txn_inline.h [hunk 2]
```diff
@@ -98,6 +98,32 @@ __wt_txn_op_set_recno(WT_SESSION_IMPL *session, uint64_t recno)
...
+static WT_INLINE bool
+__txn_op_need_set_key(WT_TXN *txn, WT_TXN_OP *op)
+{
+    /*
+     * We save the key for resolving the prepared updates. However, if we have already set the
+     * commit timestamp, the transaction cannot be prepared. Therefore, no need to save the key.
+     */
+    if (F_ISSET(txn, WT_TXN_HAS_TS_COMMIT))
+        return (false);
+
```
Comment (luke-pearson) [inline src/include/txn_inline.h:114]: Thanks for moving this into its own function. Can we add some clarifying comments?
"Prepared updates cannot exist in the history store."
"The metadata file cannot be prepared."
"Autocommit transactions cannot have prepare called on them."

File: src/include/txn_inline.h [hunk 3]
```diff
@@ -115,8 +141,7 @@ __wt_txn_op_set_key(WT_SESSION_IMPL *session, const WT_ITEM *key)
 
     op = txn->mod + txn->mod_count - 1;
 
-    if (WT_SESSION_IS_CHECKPOINT(session) || WT_IS_HS(op->btree->dhandle) ||
```
Comment (luke-pearson) [inline src/include/txn_inline.h:118]: What happened to the WT_SESSION_IS_CHECKPOINT part?

Comment (quchenhao) [reply src/include/txn_inline.h:118]: It is a duplicate check. As checkpoint can only add changes to metadata. We don't need that.

Comment (luke-pearson) [reply src/include/txn_inline.h:118]: Checkpoint can also add stuff to the history store, but I guess we check for that separately too.

**Review by luke-pearson — APPROVED** [review]
LGTM.

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Pull request merged**